feat(test-vectors): Test vector enumeration framework (#266)
awslabs/aws-encryption-sdk-specification: framework/test-vectors/test-vector-enumeration.md

[//]: # "Copyright Amazon.com Inc. or its affiliates. All Rights Reserved."
[//]: # "SPDX-License-Identifier: CC-BY-SA-4.0"

# Test Vector Enumeration 

## Overview

Test vectors are a suite of known value tests.
Successfully running test vectors asserts that
given some set of input configurations to encrypt and/or decrypt,
when these configurations are encrypted and/or decrypted,
then the expected result is achieved on encrypt and/or decrypt.

This document outlines a framework
to enumerate a set of pairings
of input configurations to encryption and/or decryption
and expected results of encryption and/or decryption.
The framework constructs this set of pairings
to be representative of all possible

> input configurations and expected results.

Reviewer: Lets say a little more here about the diff between "all possible" and "representative"

and test it in the test vectors.

For example, we might partition the set of all possible plaintexts
based on the plaintext length.
Our hypothetical "smaller sets" might be:
* Empty: length = 0
* Small: all plaintexts where (1 < length ≤ 10)
* Medium: all plaintexts where (10 < length ≤ 1000)
* Large: all plaintexts where (1000 < length ≤ 2^32-1)
* Largest frame: all plaintexts where (length = 2^32-1)
* Largest frame + partial frame: all plaintexts where (length = 2^32-1 + [1 .. 2^32-1))
* Two largest frames: all plaintexts where (length = 2*(2^32-1))
* Many frames: all plaintexts where (length = 2*(2^32-1) + [1 .. (2^32-1)*(2^32-3)])

(The motivation for partitioning this set based on lengths of 2^32-1 comes from the ESDK message [frame size](https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/message-format.html#body-framing).)

From each of these smaller sets, we select one (or more) 
concrete "representative" values that are tested in test vectors.
For example, a representative "small" plaintext could be `"abc"`.

This framework of representative values
reduces any impractically large set of test inputs
to a practically testable size

> while maintaining reasonable test coverage.

Reviewer: And documents our reasoning about what a representative value is

and determines whether the configuration is valid
under the conditions the rule evaluates.

For example,
the [required encryption context CMM component](complete-vectors/required-encryption-context-cmm.md#required-encryption-context-cmm-failures-on-encrypt)
defines the conditions for encryption to fail.
This can be interpreted as an evaluation rule,
and the test manifest generator should implement this
when generating test vectors using the required encryption context CMM
to determine when these vectors should fail.

These rules can be relatively informal in the spec,
but ideally should be written programmatically,
so a developer can write implement these rules in test manifest generation code.

Evaluation rules are a useful abstraction to determine a given test scenario's expected result. A list of independent rules is flexible and maintainable. A list of rules that maintain the same interface (reads input configuration; outputs success/fail) makes evaluating these rules programmable.

### Multiple errors (TODO)

An input configuration SHOULD only be deterministic of a single error.
If a single input configuration can result in multiple errors,

> it SHOULD NOT be written.

Reviewer: Hum, I like the idea, but what if we have 2 failure cases that BOTH really should fail.
Or another way, how do we build confidence on something that is an important secondary check?

Author: Plan is:
* Allow configurations with multiple independent failure reasons
* Note that the error description is just a helpful message, and may not accurate describe the error
  * This "absolves" the error message of accurately choosing one of the failure reasons
*  If we later decide to codify a concrete failure reason, we should have a new attribute separate from the error description.

For example,
if a test scenario specifies that `decrypt` provides incorrect `reproducedEncryptionContext` to a requried encryption context CMM
but also specifies a key that cannot decrypt the encrypted value,
that scenario should not be specified.

(Is the above accurate?

> Do we need some method to determine precedence of errors?

Reviewer 2: IMO, not in test vectors. This feels more like something we'd use per-language integ/functional tests for, since those are already more specific to individual runtime releases as opposed to the more general test-vectors.

Author: Yep, agreed.
Ryan and I chatted and decided this might be nice at some point in the very distant future, but is not worth investment right now.
In the meantime, we'll say that any "error description" is just a best-effort attempt to describe the error and should not be trusted.
(but if the error type is both negative-encrypt and negative-decrypt, write negative-encrypt.)
In the future, we might add some new error description field that is "trusted", but not right now.

given a representative small plaintext of `"abc"`,
if one were to change this to `"wxyz"`,
this should have no impact on test coverage.
This is still a small plaintext
according to the constraints on small plaintexts
and should be an allowed change.

On the other hand,
given a representative small plaintext of `"abc"`,
if one were to change this to `""`,
this removes coverage for small plaintexts.
This violates the constraint for small plaintexts
and should not be an allowed change.

The property that a representative value may be modified within its constraints
acts as an assertion that the representative value is not special,
and is truly "representative" of other values in its constrained set.

### Filtering input configurations

Input configurations may be "filtered"
and not be tested.

This is a mechanism to identify test cases that
may not be as important as others
and avoid writing them to a test vector manifest.

#### Motivation

Some input configurations produced as a result of the enumeration above
will not be tested
as they are not important enough
to warrant the time spent testing them.

For example,
consider all cases where a hierarchical keyring is encrypting,

> but some raw AES key is decrypting.

Reviewer 3: Nit/Issue: Technically, the a Branch Key's Material is an AES Key.
```suggestion
but some raw AES keyring is decrypting.
```

> This is expected have an [expected result](#expected-results) of `negative-decrypt`,

Reviewer 3: ```suggestion
This will have an [expected result](#expected-results) of `negative-decrypt`,
```

> since a hierarchical keyring's encrypt result is incompatible with AES decryption.

Reviewer 3: Nit/Suggestion: AES is Advanced Encryption Standard, which is an Algorithm, and is used by H-Keyring.
```suggestion
since a hierarchical keyring's encrypt result is incompatible with a AES Keyring's decryption.
```

> since a hierarchical keyring's encrypt result is incompatible with an AES meyring's decryption.

Reviewer 3: ```suggestion
since a hierarchical keyring's encrypt result is incompatible with a Raw AES Keyring's decryption.
```

This may be worth validating once with test vectors to ensure this scenario
fails with the expected result.

However, consider that the enumeration process above
would result in a large number of test scenarios

> with an encrypting hierarchical keyring and a decrypting raw AES key,

Reviewer 3: ```suggestion
with an encrypting hierarchical keyring and a decrypting raw AES keyring,
```

If one evaluation rule fails, the test scenario result should be the result of that evaluation rule.
If more than one evaluation rule fails with a different result
(i.e. input configuration can result in both `"negative-encrypt"` and `"negative-decrypt"`),
the test scenario result should be `"negative-encrypt"`.
(Encryption happens first; failure on encrypt will be encountered first.)

#### Expected result evaluation rules

An evaluation rule reads the input configuration [input configuration](#input-configuration),
and determines whether the configuration is valid
under the conditions the rule evaluates.

For example,
the [required encryption context CMM component](complete-vectors/required-encryption-context-cmm.md#required-encryption-context-cmm-failures-on-encrypt)
defines the conditions for encryption to fail.
This can be interpreted as an evaluation rule,
and the test manifest generator should implement this
when generating test vectors using the required encryption context CMM
to determine when these vectors should fail.

These rules can be relatively informal in the spec,
but ideally should be written programmatically,

> so a developer can write implement these rules in test manifest generation code.

Reviewer 3: ```suggestion
so a developer can implement these rules in test manifest generation code.
```

   and are not a priority to test at this time
   (see ["filtering input configurations"](#filtering-input-configurations))

These reasons are both motivated by test efficiency.
It is infeasible to enumerate every possible plaintext,
or test every invalid encryption configuration.
By strategically identifying test cases to enumerate,
the test vector framework employs orders of magnitude of leverage
to have comprehensive test coverage

> with a concise suite of tests.

Reviewer: Well said.
Also, it leaves open the door for someone to manually construct their own test case to validate or explore the possibilities
